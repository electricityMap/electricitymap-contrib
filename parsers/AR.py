#!/usr/bin/env python3


from datetime import datetime, timedelta
from logging import Logger, getLogger
from typing import Dict, List, Optional

import arrow
from pytz import timezone
from requests import Session

from .lib.exceptions import ParserException

# Useful links.
# https://en.wikipedia.org/wiki/Electricity_sector_in_Argentina
# https://en.wikipedia.org/wiki/List_of_power_stations_in_Argentina
# http://globalenergyobservatory.org/countryid/10#
# http://www.industcards.com/st-other-argentina.htm

# API Documentation: https://api.cammesa.com/demanda-svc/swagger-ui.html
CAMMESA_DEMANDA_ENDPOINT = (
    "https://api.cammesa.com/demanda-svc/generacion/ObtieneGeneracioEnergiaPorRegion/"
)

CAMMESA_RENEWABLES_ENDPOINT = "https://cdsrenovables.cammesa.com/exhisto/RenovablesService/GetChartTotalTRDataSource/"

CAMMESA_RENEWABLES_REGIONAL_ENDPOINT = (
    "https://cdsrenovables.cammesa.com/exhisto/RenovablesService/getCapacidadesRegiones"
)

CAMMESA_REGIONS_ENDPOINT = "https://api.cammesa.com/demanda-svc/demanda/RegionesDemanda"

CAMMESA_EXCHANGE_ENDPOINT = (
    "https://api.cammesa.com/demanda-svc/demanda/IntercambioCorredoresGeo/"
)

REGIONS = {
    "AR": "Total del SADI ",
    "AR-BAS": "Buenos Aires",
    "AR-NEA": "NEA",
    "AR-NOA": "NOA",
    "AR-CEN": "Centro",
    "AR-PAT": "Patagonia",
    "AR-LIT": "Litoral",
    "AR-COM": "Comahue",
    "AR-CUY": "Cuyo",
}

SUPPORTED_EXCHANGES = {
    "AR->CL-SEN": "ARG-CHI",
    "AR->PY": "ARG-PAR",
    "AR->UY": "ARG-URU",
    "AR-NEA->BR-S": "ARG-BRA",
    "AR-BAS->AR-COM": "PBA-COM",
    "AR-CEN->AR-COM": "CEN-COM",
    "AR-CEN->AR-NOA": "CEN-NOA",
    "AR-COM->AR-CUY": "CUY-COM",
    "AR-CEN->AR-CUY": "CEN-CUY",
    "AR-LIT->AR-NEA": "LIT-NEA",
    "AR-BAS->AR-LIT": "PBA-LIT",
    "AR-CUY->AR-NOA": "CUY-NOA",
    "AR-CEN->AR-LIT": "LIT-CEN",
    "AR-LIT->AR-NOA": "LIT-NOA",
    "AR-BAS->AR-CEN": "PBA-CEN",
    "AR-COM->AR-PAT": "COM-PAT",
    "AR-NEA->AR-NOA": "NEA-NOA",
}

EXCHANGE_DIRECTIONS = {  # directions are from first region -> second region
    "AR->CL-SEN": 180,
    "AR->PY": 45,
    "AR->UY": 0,
    "AR-NEA->BR-S": 45,
    "AR-BAS->AR-COM": 225,
    "AR-CEN->AR-COM": 270,
    "AR-CEN->AR-NOA": 135,
    "AR-COM->AR-CUY": 135,
    "AR-CEN->AR-CUY": 225,
    "AR-LIT->AR-NEA": 45,
    "AR-BAS->AR-LIT": 90,
    "AR-CUY->AR-NOA": 45,
    "AR-CEN->AR-LIT": 45,
    "AR-LIT->AR-NOA": 135,
    "AR-BAS->AR-CEN": 135,
    "AR-COM->AR-PAT": 270,
    "AR-NEA->AR-NOA": 180,
}


def fetch_production(
    zone_key="AR",
    session: Optional[Session] = None,
    target_datetime: Optional[datetime] = None,
    logger: Logger = getLogger(__name__),
) -> List[dict]:
    """Requests up to date list of production mixes (in MW) of a given region."""

    if target_datetime:
        raise NotImplementedError("This parser is not yet able to parse past dates")

    current_session = session or Session()

    non_renewables_production = non_renewables_production_mix(zone_key, current_session)
    renewables_production = renewables_production_mix(zone_key, current_session)

    full_production_list = [
        {
            "datetime": arrow.get(datetime_tz_ar).to("UTC").datetime,
            "zoneKey": zone_key,
            "production": merged_production_mix(
                non_renewables_production[datetime_tz_ar],
                renewables_production[datetime_tz_ar],
            ),
            "capacity": {},
            "storage": {},
            "source": "cammesaweb.cammesa.com",
        }
        for datetime_tz_ar in non_renewables_production
        if datetime_tz_ar in renewables_production
    ]

    return full_production_list


def merged_production_mix(non_renewables_mix: dict, renewables_mix: dict) -> dict:
    """Merges production mix data from different sources. Hydro comes from two
    different sources that are added up."""

    production_mix = {
        "biomass": renewables_mix["biomass"],
        "solar": renewables_mix["solar"],
        "wind": renewables_mix["wind"],
        "hydro": non_renewables_mix["hydro"] + renewables_mix["hydro"],
        "nuclear": non_renewables_mix["nuclear"],
        "unknown": non_renewables_mix["unknown"],
    }

    return production_mix


def renewables_production_mix(zone_key: str, session: Session) -> Dict[str, dict]:
    """Retrieves production mix for renewables using CAMMESA's API"""

    now = datetime.now(tz=timezone("America/Argentina/Buenos_Aires"))
    today = now.strftime("%d-%m-%Y")
    endpoint = CAMMESA_RENEWABLES_REGIONAL_ENDPOINT
    params = {}
    minute = now.minute
    rounded = minute - minute % 5
    time = now.replace(minute=rounded, second=0, microsecond=0) - timedelta(minutes=5)
    region_name = zone_key[3:]

    if zone_key == "AR":
        params = {"desde": today, "hasta": today}
        endpoint = CAMMESA_RENEWABLES_ENDPOINT

    renewables_response = session.get(endpoint, params=params)
    if renewables_response.status_code != 200:
        raise ParserException(
            "AR.py",
            f"Exception when fetching production for {zone_key}: error when calling renewables endpoint: [{renewables_response.status_code}]  {renewables_response.text}",
            zone_key,
        )

    production_list = renewables_response.json()
    sorted_production_list = sorted(production_list, key=lambda d: d["momento"])

    if zone_key != "AR":
        sorted_production_list = list(
            filter(lambda d: d["nemoRegion"] == region_name, production_list)
        )

    renewables_production: Dict[str, dict] = {
        (
            datetime.strptime(
                production_info["momento"], "%Y-%m-%dT%H:%M:%S.%f%z"
            ).isoformat()
            if zone_key == "AR"
            else time.isoformat()
        ): {
            "biomass": production_info["biocombustible"],
            "hydro": production_info["hidraulica"],
            "solar": production_info["fotovoltaica"],
            "wind": production_info["eolica"],
        }
        for production_info in sorted_production_list
    }
    return renewables_production


def non_renewables_production_mix(zone_key: str, session: Session) -> Dict[str, dict]:
    """Retrieves production mix for non renewables using CAMMESA's API"""

    id = get_region_id(zone_key, session)
    params = {"id_region": id}
    api_cammesa_response = session.get(CAMMESA_DEMANDA_ENDPOINT, params=params)
    if api_cammesa_response.status_code != 200:
        raise ParserException(
            "AR.py",
            f"Exception when fetching production for {zone_key}: error when calling non-renewables endpoint: [{api_cammesa_response.status_code}]  {api_cammesa_response.text}",
            zone_key,
        )

    production_list = api_cammesa_response.json()
    sorted_production_list = sorted(production_list, key=lambda d: d["fecha"])

    non_renewables_production: Dict[str, dict] = {
        datetime.strptime(
            production_info["fecha"], "%Y-%m-%dT%H:%M:%S.%f%z"
        ).isoformat(): {
            "hydro": production_info["hidraulico"],
            "nuclear": production_info["nuclear"],
            # As of 2022 thermal energy is mostly natural gas but
            # the data is not split. We put it into unknown for now.
            # More info: see page 21 in https://microfe.cammesa.com/static-content/CammesaWeb/download-manager-files/Sintesis%20Mensual/Informe%20Mensual_2021-12.pdf
            "unknown": production_info["termico"],
        }
        for production_info in sorted_production_list
    }
    return non_renewables_production


def get_region_id(zone_key, session: Session) -> int:
    """Fetches the region id for the zone that is required to get the production of that zone."""
    regions_response = session.get(CAMMESA_REGIONS_ENDPOINT)

    if regions_response.status_code != 200:
        raise ParserException(
            "AR.py",
            f"Exception when fetching regions for AR: error when calling regions endpoint: [{regions_response.status_code}]  {regions_response.text}",
            zone_key,
        )

    regions = regions_response.json()
    region_to_id = {region["nombre"]: region["id"] for region in regions}

    region_name = REGIONS[zone_key]
    return region_to_id[region_name]


def fetch_exchange(
    zone_key1: str,
    zone_key2: str,
    session: Optional[Session] = None,
    target_datetime: Optional[datetime] = None,
    logger: Logger = getLogger(__name__),
) -> dict:
    """Requests the last known power exchange (in MW) between two zones."""
    if target_datetime:
        raise NotImplementedError("This parser is not yet able to parse past dates")

    sorted_zone_keys = sorted([zone_key1, zone_key2])
    sorted_codes = "->".join(sorted_zone_keys)
    flow: Optional[float] = None
    returned_datetime: datetime
    exchange = {}

    if sorted_codes in SUPPORTED_EXCHANGES:
        current_session = session or Session()

        api_cammesa_response = current_session.get(CAMMESA_EXCHANGE_ENDPOINT)
        if api_cammesa_response.status_code != 200:
            raise ParserException(
                "AR.py",
                f"Exception when fetching exchange for {sorted_codes}: error when calling exchange endpoint: [{api_cammesa_response.status_code}]  {api_cammesa_response.text}",
                sorted_zone_keys,
            )

        exchange_name = SUPPORTED_EXCHANGES[sorted_codes]
        exchange_list = api_cammesa_response.json()
        for exchange in exchange_list["features"]:
            properties = exchange["properties"]
            if properties["nombre"] == exchange_name:
                angle_config = EXCHANGE_DIRECTIONS[sorted_codes]
                given_angle = int(properties["url"][6:])
                flow = int(properties["text"])
                if angle_config != given_angle:
                    flow = -flow
                returned_datetime = datetime.fromisoformat(
                    properties["fecha"][:-2] + ":" + properties["fecha"][-2:]
                )
                if flow is None:
                    raise ParserException(
                        "AR.py",
                        f"Failed fetching exchange for {sorted_zone_keys}",
                        sorted_codes,
                    )
                exchange = {
                    "sortedZoneKeys": sorted_codes,
                    "datetime": returned_datetime,
                    "netFlow": flow,
                    "source": "cammesaweb.cammesa.com",
                }
    else:
        raise NotImplementedError("This exchange is not currently implemented")

    return exchange


def fetch_price(
    zone_key: str = "AR",
    session: Optional[Session] = None,
    target_datetime: Optional[datetime] = None,
    logger: Logger = getLogger(__name__),
) -> dict:
    """Requests the last known power price of a given country."""

    raise NotImplementedError("Fetching the price is not currently implemented")


if __name__ == "__main__":
    """Main method, never used by the Electricity Map backend, but handy for testing."""

    print("fetch_production() ->")
    print(fetch_production())
    print("fetch_price() ->")
    print(fetch_price())
    print("fetch_exchange(AR, PY) ->")
    print(fetch_exchange("AR", "PY"))
    print("fetch_exchange(AR, UY) ->")
    print(fetch_exchange("AR", "UY"))
    print("fetch_exchange(AR, CL-SEN) ->")
    print(fetch_exchange("AR", "CL-SEN"))
