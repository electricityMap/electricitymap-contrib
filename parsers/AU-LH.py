from datetime import datetime, timezone
from logging import Logger, getLogger
from typing import Any, Dict, List, Optional, Union
import time

from electricitymap.contrib.lib.models.event_lists import (
    ExchangeList,
    PriceList,
    ProductionBreakdownList,
    TotalConsumptionList,
    TotalProductionList,
)
from electricitymap.contrib.lib.models.events import ProductionMix, StorageMix
from electricitymap.contrib.lib.types import ZoneKey
from parsers.lib.exceptions import ParserException

from requests_html import HTMLSession


PARSER_NAME = "AU-LH.py"
DATA_URL = "http://photonscada.com/data/perspective/client/LHI"

RENDER_TIMEOUT = 10 # seconds

def validate_and_clean_data(
    zone_key: ZoneKey,
    values: List[str]
) -> List[float]:
    if len(values) != 8:
        raise ParserException(PARSER_NAME, "Unexpected number of values parsed", zone_key)
    
    even_index_values = values[::2]
    if not all(val == "L" for val in even_index_values):
        raise ParserException(PARSER_NAME, "Unexpected parsed values (although normally discarded), website structure changed", zone_key)

    odd_index_values = values[1::2]
    cleaned_values = []

    for val in odd_index_values:
        # Observed units are always kW, this could be a wrong assumption
        unit_suffix = " kW"
        if not val.endswith(unit_suffix):
            raise ParserException(PARSER_NAME, f"Unexpected parsed values, check website structure or electricity units {odd_index_values}", zone_key)
        
        # strip suffix, convert to numerical value, convert units to MW
        parsed_val = float(val.rstrip(unit_suffix))/1000
        num_digits = 10
        cleaned_values.append(round(parsed_val, num_digits))

    return cleaned_values


def fetch_data(
    zone_key: ZoneKey,
    session: HTMLSession
) -> Dict[str, float]:
    res = session.get(DATA_URL)

    if not res.status_code == 200:
        raise ParserException(
            PARSER_NAME,
            f"Exception when fetching production error code: {res.status_code}",
            zone_key,
        )
    
    # render the website using request_html (chromium), need a sleep value to wait for render
    res.html.render(timeout=RENDER_TIMEOUT + 1, sleep=RENDER_TIMEOUT)

    # values in the DOM are poorly structured
    # thankfully the values we want are returned in tspans
    tspan_list = res.html.find("tspan")
    parsed_values = [tspan.text for tspan in tspan_list]

    cleaned_values = validate_and_clean_data(zone_key, parsed_values)

    return {
        "consumption": cleaned_values[0],
        "solar": cleaned_values[1],
        "battery": cleaned_values[2],
        "diesel": cleaned_values[3]
    }
    

def fetch_production(
    zone_key: ZoneKey,
    session: HTMLSession = None,
    target_datetime: Optional[datetime] = None,
    logger: Logger = getLogger(__name__),
) -> Union[List[dict], dict]:
    if target_datetime is not None:
        raise ParserException(PARSER_NAME, "This parser is not yet able to parse past dates", zone_key)
    
    is_new_session = session is None
    if is_new_session:
        session = HTMLSession()

    data = fetch_data(zone_key, session)

    production_list = ProductionBreakdownList(logger=logger)
    production_list.append(
        zoneKey=zone_key,
        datetime=datetime.now(timezone.utc),
        production=ProductionMix(
            gas=data["diesel"],
            solar=data["solar"],
        ),
        # storage=StorageMix(hydro=item["hydroStorage"] * -1),
        source=DATA_URL,
    )

    if is_new_session:
        session.close()

    return production_list.to_list()


if __name__ == "__main__":
    """Main method, never used by the Electricity Maps backend, but handy for testing."""

    print(fetch_production(ZoneKey("AU-LH")))
