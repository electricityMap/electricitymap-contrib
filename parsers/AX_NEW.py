from datetime import datetime
from logging import Logger, getLogger
from typing import Optional
from requests import Response, Session
from bs4 import BeautifulSoup
from re import findall

IFRAME_URL = "https://grafik.kraftnat.ax/grafer/tot_inm_24h_15.php"


def fetch_data(session: Session, logger: Logger):
    """Fetch data from the iFrame."""

    response: Response = session.get(IFRAME_URL)
    response.raise_for_status()

    soup = BeautifulSoup(response.text, "html.parser").find_all("script")
    time_series = findall(r"data: \[(.*?)\]", str(soup))[0]
    raw_data = findall(r"data:\[(.*?)\]", str(soup))
    data_list = []
    for time, sweden, alink, fossil, gustavs, wind, consumption in zip(
        time_series.split(","),
        raw_data[0].split(","),
        raw_data[1].split(","),
        raw_data[2].split(","),
        raw_data[3].split(","),
        raw_data[4].split(","),
        raw_data[5].split(","),
    ):
        data_list.append(
            {
                "time": str(time.replace('"', "")),
                "sweden": float(sweden),
                "alink": float(alink),
                "fossil": float(fossil),
                "gustavs": float(gustavs),
                "wind": float(wind),
                "consumption": float(consumption),
            }
        )
    return data_list


def fetch_production(
    zone_key: str = "AX",
    session: Session = Session(),
    target_datetime: Optional[datetime] = None,
    logger: Logger = getLogger(__name__),
) -> dict:
    """Fetch production from the iFrame."""

    if target_datetime is not None:
        raise NotImplementedError(
            "The datasource currently implemented is only real time"
        )

    data = fetch_data(session, logger)

    return data
