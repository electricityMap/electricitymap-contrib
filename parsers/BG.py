from datetime import datetime, timezone
from logging import Logger, getLogger

from requests import Session

from electricitymap.contrib.lib.models.event_lists import ProductionBreakdownList
from electricitymap.contrib.lib.models.events import ProductionBreakdown, ProductionMix
from electricitymap.contrib.lib.types import ZoneKey
from electricitymap.contrib.parsers.lib.exceptions import ParserException

PARSER = "BG.py"

SOURCE = "eso.bg"
SOURCE_API_URL = "http://www.eso.bg/api/rabota_na_EEC_json.php"

PRODUCTION_TYPES_BG_TO_EN = {
    "АЕЦ": "nuclear",
    "Кондензационни ТЕЦ": "coal",
    "Топлофикационни ТЕЦ": "gas",
    "Заводски ТЕЦ": "gas",
    "ВЕЦ": "hydro",
    "Малки ВЕЦ": "hydro",
    "ВяЕЦ": "wind",
    "ФЕЦ": "solar",
    "Био ТЕЦ": "biomass",
    "Био ЕЦ": "biomass",
    "Товар РБ": "consumption",
}


def _fetch_all_production_events(
    zone_key: ZoneKey,
    session: Session,
    target_datetime: datetime | None,
    logger: Logger,
) -> list[ProductionBreakdown]:
    """Fetches all production events from the backend source API."""
    if target_datetime is not None:
        raise ParserException(
            PARSER, "This parser is not yet able to parse historical data", zone_key
        )

    time = datetime.now(timezone.utc)
    response = session.get(SOURCE_API_URL)
    if not response.status_code == 200:
        raise ParserException(
            PARSER,
            f"Exception when fetching production error code: {response.status_code}: {response.text}",
            zone_key,
        )

    response_payload = response.json()
    logger.debug(f"Raw generation breakdown: {response_payload}")

    production = {}
    for header, value in response_payload:
        for production_type_bg in PRODUCTION_TYPES_BG_TO_EN:
            if header.startswith(production_type_bg):
                production_type_en = PRODUCTION_TYPES_BG_TO_EN[production_type_bg]
                production[production_type_en] = (
                    production.get(production_type_en, 0.0) + value
                )

    event = ProductionBreakdown.create(
        logger=logger,
        zoneKey=zone_key,
        datetime=time,
        source=SOURCE,
        production=ProductionMix(
            biomass=production.get("biomass"),
            coal=production.get("coal"),
            gas=production.get("gas"),
            geothermal=production.get("geothermal"),
            hydro=production.get("hydro"),
            nuclear=production.get("nuclear"),
            oil=production.get("oil"),
            solar=production.get("solar"),
            unknown=production.get("unknown"),
            wind=production.get("wind"),
        ),
        storage=None,
    )
    return [event]


def fetch_production(
    zone_key: ZoneKey = ZoneKey("BG"),
    session: Session | None = None,
    target_datetime: datetime | None = None,
    logger: Logger = getLogger(__name__),
) -> list[dict]:
    session = session or Session()
    all_production_events = _fetch_all_production_events(
        zone_key=zone_key,
        session=session,
        target_datetime=target_datetime,
        logger=logger,
    )
    production_list = ProductionBreakdownList(logger)
    for event in all_production_events:
        production_list.append(
            zoneKey=event.zoneKey,
            datetime=event.datetime,
            source=event.source,
            production=event.production,
            storage=event.storage,
            sourceType=event.sourceType,
        )
    return production_list.to_list()
