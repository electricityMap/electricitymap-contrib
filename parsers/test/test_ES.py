from datetime import datetime, timezone
from unittest import TestCase, main

from mock import patch
from ree import ElHierro, Formentera, Mallorca, Response
from requests import Session

from electricitymap.contrib.lib.types import ZoneKey
from parsers import ES


class TestES(TestCase):
    def setUp(self):
        self.session = Session()

        first_request = Response(1687192328)
        first_request.link["pe_ma"] = 50
        first_request.link["ib_fo"] = 10

        second_request = Response(1687192357)
        second_request.link["pe_ma"] = 50
        second_request.link["ib_fo"] = 10

        self.mocked_responses = [first_request, second_request]

    ### El Hierro
    # Consumption
    @patch.object(ElHierro, "get_all")
    def test_fetch_consumption(self, mocked_get_all):
        mocked_get_all.return_value = self.mocked_responses
        data_list = ES.fetch_consumption(ZoneKey("ES-CN-HI"), self.session)
        self.assertEqual(len(data_list), 2)
        for data in data_list:
            self.assertEqual(data["zoneKey"], "ES-CN-HI")
            self.assertEqual(data["source"], "demanda.ree.es")
            self.assertTrue(isinstance(data["datetime"], datetime))
            self.assertIsNotNone(data["consumption"])
        self.assertEqual(
            data_list[0]["datetime"],
            datetime.fromtimestamp(1687192328).astimezone(timezone.utc),
        )

    # Production
    @patch.object(ElHierro, "get_all")
    def test_fetch_production_el_hierro(self, mocked_get_all):
        mocked_get_all.return_value = [
            Response(1687192328, 10.0, 10, 5, 13, 3, 4, 3.5, 10, 5, 0, 3),
        ]
        data_list = ES.fetch_production(ZoneKey("ES-CN-HI"), self.session)
        for data in data_list:
            self.assertEqual(data["zoneKey"], "ES-CN-HI")
            self.assertEqual(data["source"], "demanda.ree.es")
            self.assertTrue(isinstance(data["datetime"], datetime))
            self.assertIsNotNone(data["production"])
            self.assertIsNotNone(data["storage"])
            self.assertEqual(
                data["datetime"],
                datetime.utcfromtimestamp(1687192328).replace(tzinfo=timezone.utc),
            )
            self.assertEqual(data["production"]["gas"], 0)
            self.assertEqual(data["production"]["oil"], 22)
            self.assertEqual(data["production"]["hydro"], 0)
            self.assertEqual(data["production"]["solar"], 3.5)
            self.assertEqual(data["production"]["wind"], 13)
            self.assertEqual(data["production"]["coal"], 5)
            self.assertEqual(data["production"]["unknown"], 3)
            self.assertEqual(data["storage"]["hydro"], -10)

    ### Formentera
    # Production
    @patch.object(Formentera, "get_all")
    def test_fetch_production_ES_IB_FO(self, mocked_get_all):
        mocked_get_all.return_value = [
            Response(
                timestamp=1687192328,
                demand=10.0,
                diesel=0,
                gas=5,
                wind=13,
                combined=3,
                vapor=4,
                solar=3.5,
                hydraulic=10,
                carbon=5,
                other=3,
            ),
            Response(
                timestamp=1687192328,
                demand=10.0,
                diesel=-10,  # Test negative value
                gas=5,
                wind=13,
                combined=3,
                vapor=4,
                solar=3.5,
                hydraulic=10,
                carbon=5,
                other=3,
            ),
        ]
        data_list = ES.fetch_production(ZoneKey("ES-IB-FO"), self.session)
        self.assertEqual(len(data_list), 2)
        for data in data_list:
            self.assertEqual(data["zoneKey"], "ES-IB-FO")
            self.assertEqual(data["source"], "demanda.ree.es")
            self.assertTrue(isinstance(data["datetime"], datetime))
            self.assertIsNot(data["production"], {})
            self.assertEqual(
                data["datetime"],
                datetime.utcfromtimestamp(1687192328).replace(tzinfo=timezone.utc),
            )
            self.assertEqual(data["production"]["gas"], 8)
            self.assertEqual(data["production"]["oil"], 4)
            self.assertEqual(data["production"]["hydro"], 10)
            self.assertEqual(data["production"]["solar"], 3.5)
            self.assertEqual(data["production"]["wind"], 13)
            self.assertEqual(data["production"]["coal"], 5)
            self.assertEqual(data["production"]["unknown"], 3)

    # Exchange
    @patch.object(Formentera, "get_all")
    def test_fetch_exchange_ES_IB_FO(self, mocked_get_all):
        mocked_get_all.return_value = self.mocked_responses
        data_list = ES.fetch_exchange(
            ZoneKey("ES-IB-FO"), ZoneKey("ES-IB-IZ"), self.session
        )
        self.assertEqual(len(data_list), 2)
        for data in data_list:
            self.assertEqual(data["sortedZoneKeys"], "ES-IB-FO->ES-IB-IZ")
            self.assertEqual(data["source"], "demanda.ree.es")
            self.assertIsNotNone(data["netFlow"])
            self.assertEqual(data["netFlow"], -10.0)
            self.assertTrue(isinstance(data["datetime"], datetime))
        self.assertEqual(
            data_list[0]["datetime"],
            datetime.utcfromtimestamp(1687192328).replace(tzinfo=timezone.utc),
        )

    ### Mallorca
    # Exchange
    @patch.object(Mallorca, "get_all")
    def test_fetch_exchange(self, mocked_get_all):
        mocked_get_all.return_value = self.mocked_responses
        data_list = ES.fetch_exchange(ZoneKey("ES"), ZoneKey("ES-IB-MA"), self.session)
        self.assertEqual(len(data_list), 2)
        for data in data_list:
            self.assertEqual(data["sortedZoneKeys"], "ES->ES-IB-MA")
            self.assertEqual(data["source"], "demanda.ree.es")
            self.assertIsNotNone(data["netFlow"])
            self.assertEqual(data["netFlow"], 50.0)
            self.assertTrue(isinstance(data["datetime"], datetime))
        self.assertEqual(
            data_list[0]["datetime"],
            datetime.utcfromtimestamp(1687192328).replace(tzinfo=timezone.utc),
        )


if __name__ == "__main__":
    main()
