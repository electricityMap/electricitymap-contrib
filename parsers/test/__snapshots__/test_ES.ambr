# serializer version: 1
# name: test_fetch_consumption
  list([
    dict({
      'consumption': 5.5,
      'datetime': datetime.datetime(2023, 9, 4, 0, 55, tzinfo=zoneinfo.ZoneInfo(key='Atlantic/Canary')),
      'source': 'demanda.ree.es',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'zoneKey': 'ES-CN-HI',
    }),
  ])
# ---
# name: test_fetch_exchange
  list([
    dict({
      'datetime': datetime.datetime(2023, 9, 3, 19, 5, tzinfo=zoneinfo.ZoneInfo(key='Europe/Madrid')),
      'netFlow': 309.8,
      'sortedZoneKeys': 'ES->ES-IB-MA',
      'source': 'demanda.ree.es',
      'sourceType': <EventSourceType.measured: 'measured'>,
    }),
  ])
# ---
# name: test_fetch_production
  list([
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2023, 9, 2, 21, 0, tzinfo=zoneinfo.ZoneInfo(key='Europe/Madrid')),
      'production': dict({
        'biomass': 0.0,
        'coal': 0.0,
        'gas': 0.0,
        'oil': 81.9,
        'solar': 1.1,
        'unknown': 0.0,
        'wind': 0.2,
      }),
      'source': 'demanda.ree.es',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
      }),
      'zoneKey': 'ES-IB-ME',
    }),
  ])
# ---
# name: test_fetch_production_storage
  list([
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2023, 9, 4, 0, 55, tzinfo=zoneinfo.ZoneInfo(key='Atlantic/Canary')),
      'production': dict({
        'oil': 3.2,
        'solar': 0.0,
        'wind': 3.0,
      }),
      'source': 'demanda.ree.es',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
        'hydro': 0.5,
      }),
      'zoneKey': 'ES-CN-HI',
    }),
  ])
# ---
# name: test_production_DST_CN
  list([
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2021, 10, 31, 1, 55, tzinfo=datetime.timezone(datetime.timedelta(seconds=3600))),
      'production': dict({
        'oil': 2.8,
        'solar': 0.0,
        'wind': 3.3,
      }),
      'source': 'demanda.ree.es',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
        'hydro': 1.2,
      }),
      'zoneKey': 'ES-CN-HI',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2021, 10, 31, 1, 0, tzinfo=datetime.timezone.utc),
      'production': dict({
        'oil': 2.5,
        'solar': 0.0,
        'wind': 3.2,
      }),
      'source': 'demanda.ree.es',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
        'hydro': 1.2,
      }),
      'zoneKey': 'ES-CN-HI',
    }),
  ])
# ---
# name: test_production_DST_IB
  list([
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2020, 10, 25, 2, 55, tzinfo=datetime.timezone(datetime.timedelta(seconds=7200))),
      'production': dict({
        'biomass': 19.2,
        'coal': 73.3,
        'gas': 127.1,
        'oil': 0.0,
        'solar': 0.0,
        'unknown': 2.9,
        'wind': 0.0,
      }),
      'source': 'demanda.ree.es',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
      }),
      'zoneKey': 'ES-IB-MA',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2020, 10, 25, 2, 0, tzinfo=datetime.timezone(datetime.timedelta(seconds=3600))),
      'production': dict({
        'biomass': 18.5,
        'coal': 73.3,
        'gas': 128.0,
        'oil': 0.0,
        'solar': 0.0,
        'unknown': 2.9,
        'wind': 0.0,
      }),
      'source': 'demanda.ree.es',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
      }),
      'zoneKey': 'ES-IB-MA',
    }),
  ])
# ---
